### ISSUE TITLE
Add GOP length to VideoEncoderConfig

### ISSUE THREAD

**Issue by HuaChunbo** [w3c/webcodecs#444]
GOP (group of picture) length indicates the distance between two encoded key frames, which affects the encoded bitstream quite a lot, like the bitrate etc. Providing GOP length attribute in VideoEncoderConfig will help user to control the output encoded bitstream better.

**Comment by aboba** [comment]
[VideoEncoderEncodeOptions](https://w3c.github.io/webcodecs/#dictdef-videoencoderencodeoptions) allows the application to determine the frame type.

**Comment by HuaChunbo** [comment]
GOP length is more like the behavior of encoder itself from [VideoEncoderConfig](https://w3c.github.io/webcodecs/#dictdef-videoencoderconfig) possibly, in order to periodically generate output bitstream with specified frame length. At the same time, [VideoEncoderEncodeOptions](https://w3c.github.io/webcodecs/#dictdef-videoencoderencodeoptions) is used to make sure encoder generate a key frame for a specific input frame.
So basically the usage [VideoEncoderEncodeOptions](https://w3c.github.io/webcodecs/#dictdef-videoencoderencodeoptions) for key frame generation will only shorten the encoder's internal key frame period, not set any GOP length.

**Comment by Satkhiramch** [comment]
Pfizer bioNTech

**Comment by Djuffin** [comment]
What do you mean by "shorten the encoder's internal key frame period" ?

Key frame distance might be a convenient option, but I kinda agree with Bernard that it's trivial to implement on application level.

**Comment by HuaChunbo** [comment]
My point is that encoder can have its own GOP length by default, like for some hardware encoder provider, it is 100 for example, which is no good for real-time scenario like conference. For now, WebCodecs is using 10000 as the GOP length configuration for all. I think it is better for WebCodecs users to define the output bitstream with proper GOP length per their requirement.

**Comment by chcunningham** [comment]
Authors may use VideoEncoderEncodeOptions to require a keyframe, but they can't use it to prevent a keyframe. Chrome's default of 10000 isn't required by the spec. Spec should clarify what the default interval is to avoid having implementations choosing an interval like 100 that breaks RTC scenarios. Alternatively, spec could add a gop length config and simply say the default is implementer defined. 

config.gopLength semantics are another thing to consider. Would this be a max gopLength, still leaving the encoder room to send keyframes more frequently as it chooses? I believe that's how it works for Chrome's [Vpx](https://source.chromium.org/chromium/chromium/src/+/main:media/video/vpx_video_encoder.cc;drc=86223e421af31a52ac959ae84b71405a64a9092b;l=111) and [Av1](https://source.chromium.org/chromium/chromium/src/+/main:media/video/av1_video_encoder.cc;l=94?q=keyframe_interval%20f:encode%20f:cc%20-f:test%20-f:third&ss=chromium%2Fchromium%2Fsrc) encoders. I'm not sure how [this knob in openH264 works](https://source.chromium.org/chromium/chromium/src/+/main:media/video/openh264_video_encoder.cc;drc=86223e421af31a52ac959ae84b71405a64a9092b;l=39). @Djuffin, thoughts? Do you know the hw encode story here?

**Comment by dalecurtis** [comment]
Since this is platform specific it's not really something we can add much to in the spec beyond a note. If you have proposed text we can consider adding something.
